core/iwasm/fast-jit/cg/x86-64/jit_codegen_x86_64.cpp

```diff
@@ -6242,59 +6388,1271 @@ cast_r_f64_to_r_i64(x86::Assembler &a, int32 reg_no_dst, int32 reg_no_src)
+ * @param reg_no_base the no of register that stores the base address
+ *        of src&dst memory(first operand&store back location)
+ * @param offset the offset address of the memory
+ * @return true if success, false otherwise
+ */
+static bool
+at_rmw_xchg_r_base_r_offset_imm(x86::Assembler &a, uint32 bytes_dst,
+                                uint32 kind_dst, int32 reg_no_dst,
+                                int32 reg_no_src, int32 reg_no_base,
+                                int32 offset)
+{
+    x86::Mem m(regs_i64[reg_no_base], offset, bytes_dst);
+    return xchg_r_to_m(a, bytes_dst, kind_dst, m, reg_no_src)
+           && extend_r_to_r_kind(a, bytes_dst, kind_dst, reg_no_src,
+                                 reg_no_dst);
+}
+
+/**
+ * Encode atomic rmw xchg: load value into a register from memory
+ * with reg base and reg offset, exchange loaded value with reg data, store back
+ *
+ * @param a the assembler to emit the code
+ * @param bytes_dst the bytes number of the data to actual operated on(load,
+ * compare, replacement) could be 1(byte), 2(short), 4(int32), 8(int64)
+ * @param reg_no_dst the no of register that stores the returned value
+ * @param reg_no_src the no of register store the src data(second operand)
+ * @param reg_no_base the no of register that stores the base address
+ *        of src&dst memory(first operand&store back)
+ * @param reg_no_offset the no of register that stores the offset of the memory
+ * @return true if success, false otherwise
+ */
+static bool
+at_rmw_xchg_r_base_r_offset_r(x86::Assembler &a, uint32 bytes_dst,
+                              uint32 kind_dst, int32 reg_no_dst,
+                              int32 reg_no_src, int32 reg_no_base,
+                              int32 reg_no_offset)
+{
+    x86::Mem m(regs_i64[reg_no_base], regs_i64[reg_no_offset], 0, 0, bytes_dst);
+    return xchg_r_to_m(a, bytes_dst, kind_dst, m, reg_no_src)
+           && extend_r_to_r_kind(a, bytes_dst, kind_dst, reg_no_src,
+                                 reg_no_dst);
+}
+
+/**
+ * Encode insn rmw logical operation: generate a loop to make sure it's atomic
+ * @param bin_op the operation, can be and/or/xor
+ * @param kind the data kind, can only be I32 or I64
+ * @param bytes_dst the byte number of dst data
+ */
+#define AT_RMW_LOGICAL_LOOP(bin_op, kind, bytes_dst)                           \
+    do {                                                                       \
+        bh_assert((kind_dst == JIT_REG_KIND_I32 && bytes_dst <= 4)             \
+                  || kind_dst == JIT_REG_KIND_I64);                            \
+        bh_assert(reg_no_src < 16 && reg_no_dst < 16);                         \
+        /* read original value in memory(operand 1) to rax(expected) */        \
+        mov_m_to_r(a, bytes_dst, kind_dst, false, REG_RAX_IDX, m_dst);         \
+        Label loop = a.newLabel();                                             \
```

Comment:
Should check return value, if error, return false

Reply:
And not sure whether using label is a good choice, since the compiled code isn't the final code, it will be copied into the actual machine code allocated:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/fast-jit/cg/x86-64/jit_codegen_x86_64.cpp#L6746-L6755

Reply:
That's a good point. I tested the atomicity and correctness of  `xor opcode`, and the result seems to be fine. I will use check document and us gdb to further check the correctness of this usage